scripts/merge-script/index.ts

```diff
@@ -0,0 +1,333 @@
+                                      PullRequest): Promise<PullRequestFailure|null> {
+    let previousBranch: string|null = null;
+    try {
+      previousBranch = this._git(['rev-parse', '--abbrev-ref', 'HEAD']).stdout.trim();
+      // Fetch the pull request and store it into our temporary pull request head branch.
+      this._git(['fetch', '-f', this._repoGitUrl, `pull/${prNumber}/head:${PR_HEAD_NAME}`]);
+
+      // In case a required base is specified for this pull request, check if the pull
+      // request contains the given commit. If not, return a pull request failure. This
+      // check is useful for enforcing that PRs are rebased on top of a given commit. e.g.
+      // a commit that changes the codeowner ship validation. PRs which are not rebased
+      // could bypass new codeowner ship rules.
+      if (requiredBaseSha && !this._hasCommit(PR_HEAD_NAME, requiredBaseSha)) {
+        return {prNumber, status: PullRequestStatus.NEEDS_REBASE, context: requiredBaseSha};
+      }
+
+      // Create a new branch for the base of the pull request. This branch is created
+      // so that we can easily compute a revision that contains all pull request commits.
+      this._git(['branch', '-f', PR_BASE_NAME, `${PR_HEAD_NAME}~${commitsAmount}`]);
+
+      // Update pull requests commits to reference the pull request. This matches what
+      // Github does when pull requests are merged through the Web UI. The motivation is
+      // that it should be easy to determine which pull request contained a given commit.
+      this._git([
+        'filter-branch', '-f', '--msg-filter', `${MSG_FILTER_SCRIPT} ${prNumber}`,
+        `${PR_BASE_NAME}..${PR_HEAD_NAME}`
+      ]);
+
+      // Non-interactive rebase the pull request to squash all fixup or squash commits.
+      this._git(['rebase', '--autosquash', PR_BASE_NAME, PR_HEAD_NAME]);
+
+      // Cherry-pick the pull request into all determined target branches.
+      for (const branch of targetBranches) {
+        const localTargetBranch = this._getLocalTargetBranchName(branch);
+        // Checkout the local target branch.
+        this._git(['checkout', localTargetBranch]);
+        // Cherry-pick the pull request commits into the target branch.
+        if (this._gitGraceful(['cherry-pick', `${PR_BASE_NAME}..${PR_HEAD_NAME}`]).status !== 0) {
+          // Abort the failed cherry-pick. We do this because Git persists the failed
+          // cherry-pick state globally in the repository. This could prevent future
+          // pull request merges as a Git thinks a cherry-pick is still in progress.
+          this._gitGraceful(['cherry-pick', '--abort']);
+          return {prNumber, status: PullRequestStatus.CHERRY_PICK_CONFLICTS, context: branch};
+        }
+      }
+      return null;
+    } catch {
+      return {prNumber, status: PullRequestStatus.UNKNOWN_MERGE_ERROR};
+    } finally {
+      // Always try to restore the branch if possible. We don't want to leave the
```

Comment:
Additional to always trying to restore the branch we came from, we should also clean up all of the branches we created correct?